Migrate TabInCollectionViewHolder to Compose

Mugurell opened the pull request:
Draft for now. Still a bit of cleanup to do.


### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

Mugurell commented:
Rebasing to fix conflicts.

gabrielluong commented:
Extracted out the WidgetSiteItem PR into https://github.com/mozilla-mobile/fenix/pull/24933

Mugurell commented:
Rebased on top of latest `main`.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/home/collections/TabInCollection.kt:
```diff
@@ -0,0 +1,300 @@
+        onRemove(true)
+    }
+
+    SwipeToDismiss(
```
I am curious if you used any particular resources/docs for this. Just want to be looking at the same resources while reviewing this.

Mugurell replied:
There the [official documentation](https://developer.android.com/reference/kotlin/androidx/compose/material/package-summary#SwipeToDismiss(androidx.compose.material.DismissState,androidx.compose.ui.Modifier,kotlin.collections.Set,kotlin.Function1,kotlin.Function1,kotlin.Function1)) based on which I tried to obtain the exact same result as we have now. (The delete icons from the background which appear selectively based on the swipe direction needed special attention)

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/collections/TabInCollection.kt:
```diff
@@ -0,0 +1,300 @@
+/**
+ * Clips the Composable this applies to such that it cannot draw content / shadows outside it's top bound.
+ */
+private fun Modifier.clipTop() = this.then(
```
do we foresee this being used elsewhere? If so, would it make sense to have this live in the `Modifier` extension file?

gabrielluong replied:
We probably don't want to keep this around longer than we have to. The plan is to join all the collection composables under a single Card. See https://github.com/mozilla-mobile/fenix/pull/24501#discussion_r869410027

Mugurell replied:
Thank you!
There are a number of bugs with the current homescreen collections (also filed recently [some](https://github.com/mozilla-mobile/fenix/issues?q=is%3Aopen+is%3Aissue+author%3AMugurell+label%3AFeature%3ACollections))  I'd say mostly stemming from the cumbersome data flow of this feature. It would need to be migrated to our MVI store and then one composable observing said store would have an easier time of syncing everything.
This patch only ment to do a 1-1 migration from the old View based system to Compose. Same bugs should still exist and work exactly as before 💪.
The above modifier was needed for a new situation in which it seems like two separate composables will show a bit of a shadow in between. May be a bug in Compose but it should also be resolved by having everything in just on ViewHolder. It does seems like a bit of a smell and I'd agree we don't want to keep it around any longer than is expressly needed.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/collections/TabInCollection.kt:
```diff
@@ -0,0 +1,300 @@
+            FaviconListItem(
+                label = tab.title,
+                description = shortenUrl(tab.url),
+                onClick = { onClick() },
```
```suggestion
                onClick = onClick,
```

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/collections/TabInCollection.kt:
```diff
@@ -0,0 +1,300 @@
+        Card(
+            modifier = clippingModifier
+                .fillMaxWidth()
+                .defaultMinSize(minHeight = 56.dp),
```
`defaultMinSize` may not be necessary, since `FaviconListItem` has this already set.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/collections/TabInCollection.kt:
```diff
@@ -0,0 +1,300 @@
+                    dismissDirection = StartToEnd,
+                    isLastInCollection = false,
+                )
+            }
```
nit: missing a couple new line between Compose blocks in the previews

Mugurell replied:
Was intended on my part. Wanted to group in code what's shown in preview.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/collections/TabInCollection.kt:
```diff
@@ -0,0 +1,300 @@
+}
+
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+@Preview(uiMode = Configuration.UI_MODE_NIGHT_YES)
```
```suggestion
@Composable
@Preview(uiMode = Configuration.UI_MODE_NIGHT_YES)
@Preview(uiMode = Configuration.UI_MODE_NIGHT_NO)
```

instead of doing two duplicate preview blocks for light/dark, it's possible to do two consecutive preview calls.

Mugurell replied:
Didn't know that. Awesome tip, thank you!

Mugurell replied:
Tried it now exactly as written here and I see two previews with both using light mode though even after "Build Refresh".
Could be needing a newer Android Studio?
Will remove one version nonetheless.

MozillaNoah replied:
Is the theme still being set to light in the `FirefoxTheme` initialization?

Mugurell replied:
😅 That was it.
Removing it I see errors in the preview rendering:
> java.lang.ClassCastException: class com.android.layoutlib.bridge.android.BridgeContext cannot be cast to class org.mozilla.fenix.FenixApplication (com.android.layoutlib.bridge.android.BridgeContext is in unnamed module of loader com.intellij.ide.plugins.cl.PluginClassLoader @195c4744; org.mozilla.fenix.FenixApplication is in unnamed module of loader org.jetbrains.android.uipreview.ModuleClassLoader @283fdca)   at org.mozilla.fenix.ext.ContextKt.getApplication(Context.kt:29)   at org.mozilla.fenix.ext.ContextKt.getComponents(Context.kt:35)   at org.mozilla.fenix.ext.ContextKt.settings(Context.kt:57)   at org.mozilla.fenix.theme.Theme$Companion.getTheme(FirefoxTheme.kt:40)

Checked the codebase, seems like this cool double preview of the themed composable isn't used anywhere else so maybe we can't use it just yet :-(.

MozillaNoah replied:
It's because of a limitation in `FirefoxTheme` currently, so the initialization needs to be:

`FirefoxTheme(theme = Theme.getTheme(isPrivate = false)) {`

Mugurell replied:
Worked! Thank you!

Mugurell replied:
I think I remember that limitation being because of the compose version used? And I think now that can be avoided?

MozillaNoah reviewed (commented):
initial pass of comments

Mugurell commented:
Rebased and merged everything in one commit as gabrielluong requested.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/home/collections/TabInCollection.kt:
```diff
@@ -0,0 +1,262 @@
+ * @see toShortUrl
+ */
+@Composable
+private fun shortenUrl(url: String): String {
```
We landed an extension function for `TabSessionState` that does the same thing in https://github.com/mozilla-mobile/fenix/pull/25242. It's kinda unfortunate, but maybe we should've actually add a `String` extension function instead of `TabSessionState`

Mugurell replied:
Not sure I have to do anything about this now.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/home/collections/TabInCollection.kt:
```diff
@@ -0,0 +1,262 @@
+@Composable
+private fun DismissedTabBackground(
+    dismissDirection: DismissDirection?,
+    isLastInCollection: Boolean,
```
I also have similar feelings for `TabInCollection`

Mugurell replied:
Seems like the decision is about who should control what shape the background looks like for a dismissed tab.

With all the other values related to size, colors and imagery controlled from this composable and no way for it to prevent clients passing a circle for example and breaking it seems like it is better for this composable to be in full control of the shape also and for it to be left to contain all rules on how to adapt a bit of state to show this particular piece of UI.
If others want to pass another shape probably the usecase is different and other properties (sizes, colors, imagery) have to be updated also.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/home/collections/TabInCollectionViewHolder.kt:
```diff
@@ -0,0 +1,91 @@
+
+        tabInfo.value.tab?.let { validTab ->
+            tabInfo.value.collection?.let { validCollection ->
+                val horizontalPadding =
+                    composeView.resources.getDimensionPixelSize(R.dimen.home_item_horizontal_margin)
+                composeView.setPadding(horizontalPadding, 0, horizontalPadding, 0)
```
Let's move this into the `init {}` since that's what we do everywhere.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/home/collections/TabInCollectionViewHolder.kt:
```diff
@@ -0,0 +1,91 @@
+    override fun Content() {
+        val tabInfo = remember { mutableStateOf(tabData) }
+
+        tabInfo.value.tab?.let { validTab ->
```
```suggestion
        tabInfo.value.tab?.let { tab ->
```

gabrielluong commented on app/src/main/java/org/mozilla/fenix/home/collections/TabInCollectionViewHolder.kt:
```diff
@@ -0,0 +1,91 @@
+        val tabInfo = remember { mutableStateOf(tabData) }
+
+        tabInfo.value.tab?.let { validTab ->
+            tabInfo.value.collection?.let { validCollection ->
```
```suggestion
            tabInfo.value.collection?.let { collection ->
```

gabrielluong commented on app/src/main/java/org/mozilla/fenix/home/collections/TabInCollectionViewHolder.kt:
```diff
@@ -0,0 +1,91 @@
+
+                TabInCollection(
+                    tab = validTab,
+                    isLastInCollection = tabInfo.value.isLastInCollection,
```
Maybe we just determine the `shape` here?

Mugurell replied:
As above allowing to pass shapes leads to brittle composables that can easily be broken by clients passing a "wrong" shape.
I feel like this is a good indication that we shouldn't. Only pass indications / state on what needs to be shown and leave this Composable applying it's business rules on adapting that and in this ensuring consistency with the rest of "hardcoded" properties it uses.

Mugurell replied:
If this exact piece of UX (represented by this Composable) will be needed in other places later maybe we can just "promote" it to the `compose` package and change the naming of the `isLastInCollection` parameter but otherwise if there's any need for other shapes then other properties will probably have to be changed also. 

(Still thinking about passing a circle or some kind of an irregular shape. Shape's middle name is "Danger".)

I guess most we could do is allowing to pass a "cornerRadius" value which we can confine in between safe values. If this will be needed later.

Mugurell commented on app/src/main/java/org/mozilla/fenix/home/collections/TabInCollection.kt:
```diff
@@ -0,0 +1,300 @@
+ */
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+fun TabInCollection(
```
Went with `CollectionItem`.

Mugurell commented:
Thank you for helping with this @gabrielluong !
I've rebased and addressed some of the above comments (marked now as resolved).
Open to do more changes if you feel strongly about them.

gabrielluong commented:
> Thank you for helping with this @gabrielluong ! I've rebased and addressed some of the above comments (marked now as resolved). Open to do more changes if you feel strongly about them.

I think it's fine to land as is
